In [1]:
import pandas as pd
from pandas import DataFrame,Series
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import math
import random
import statsmodels.stats.weightstats as sw

In [2]:
from pylab import mpl
mpl.rcParams['font.sans-serif'] = ['Microsoft YaHei'] # 指定默认字体
mpl.rcParams['axes.unicode_minus'] = False # 解决保存图像是负号'-'显示为方块的问题

# 例题7.2

设高、中、低管理者评分的均值为u1、u2、u3

## 提出假设

H0：u1=u2=u3

H1：u1、u2、u3不全相等

## 计算均值

In [3]:
x1 = (7+7+8+7+9)/5
x2 = (8+9+8+10+9+10+8)/7
x3 = (5+6+5+7+4+8)/6
x = (x1+x2+x3)/3
print("x1:%f x2:%f x3:%f\nx:%f"%(x1,x2,x3,x))

x1:7.600000 x2:8.857143 x3:5.833333
x:7.430159


## 计算SSA和SSE

In [4]:
ssa = ((x1-x)**2)*5+((x2-x)**2)*7+((x3-x)**2)*6
sse = (((7-x1)**2)+((7-x1)**2)+((8-x1)**2)+((7-x1)**2)+((9-x1)**2)) + (((8-x2)**2)+((9-x2)**2)+((8-x2)**2)+((10-x2)**2)+((9-x2)**2))+(((10-x2)**2)+((8-x2)**2)) + ((((5-x3)**2)+((6-x3)**2)+((5-x3)**2)+((7-x3)**2)+((4-x3)**2))+((8-x3)**2))                                                        
print("ssa:%f\nsse:%f"%(ssa,sse))

ssa:29.697324
sse:18.890476


## 计算检验统计量F

In [5]:
dfa = 2
dfe = 15
msa = ssa/dfa
mse = sse/dfe
f = msa/mse
# F~F(2,15)
F = stats.f.ppf(0.95,dfn=3,dfd=20)
print("dfa:%f dfe:%f\nmsa:%f mse:%f\nf:%f\nF临界值:%f"%(dfa,dfe,msa,mse,f,F))

dfa:2.000000 dfe:15.000000
msa:14.848662 mse:1.259365
f:11.790594
F临界值:3.098391


单因素方差分析结果：

| 误差来源 | 离差平方和SS | 自由度df | 均方MS    | 检验统计量F | P值      | F临界值  |
| -------- | ------------ | -------- | --------- | ----------- | -------- | -------- |
| 组件     | 29.697324    | 2        | 14.848662 | 11.790594      | 0.000849 | 3.098391 |
| 组内     | 18.890476    | 15       | 1.259365  |             |          |          |
| 合计     | 48.587800    | 17       |           |             |          |          |

## 决策

因为统计检测量F（11.790594）大于 F临界值（3.098391），所以因拒绝原假设

# 例题7.5

## 提出假设

对列因素A：

H0:u1=u2=u3=u4=u5

H1:u1、u2、u3、u4、u5不全相等

对列因素B：

H0:u1=u2=u3=u4

H1:u1、u2、u3、u4、u5不全相等

## 计算各离差平方和

In [6]:
ssa=4*(((20.5-14.1)**2)+((10-14.1)**2)+((16.5-14.1)**2)+((11.5-14.1)**2)+((12-14.1)**2))
ssb=5*(((15.2-14.1)**2)+((14-14.1)**2)+((16.8-14.1)**2)+((10.4-14.1)**2))
sst=643.8 # 式子太长了...
sse=sst-ssa-ssb
print("ssa:%f\nssb:%f\nsst:%f\nsse:%f"%(ssa,ssb,sst,sse))

ssa:298.800000
ssb:111.000000
sst:643.800000
sse:234.000000


In [7]:
msa=ssa/(5-1)
msb=ssb/(4-1)
mse=sse/(20-5-4+1)
fa=msa/mse
fb=msb/mse
FA=stats.f.ppf(0.95,dfn=4,dfd=12)
FB=stats.f.ppf(0.95,dfn=3,dfd=12)
print("msa:%f msb:%f\nmse:%f\nfa:%f fb:%f\nFA临界值:%f FB临界值:%f"%(msa,msb,mse,fa,fb,FA,FB))

msa:74.700000 msb:37.000000
mse:19.500000
fa:3.830769 fb:1.897436
FA临界值:3.259167 FB临界值:3.490295


单因素方差分析结果：

| 误差来源 | 离差平方和SS | 自由度df | 均方MS    | 检验统计量F | P值      | F临界值  |
| -------- | ------------ | -------- | --------- | ----------- | -------- | -------- |
| 因素A     | 298.800000    | 4       | 74.700000 | 3.830769      | 0.031324 | 3.259167 |
| 因素B     | 111.000000    | 3      | 37.000000  |     1.897436        |   0.183889       |   3.490295       |
| 随机因素     | 234.000000   | 12       |   |             |          |          |
| 合计     | 48.587800    | 19      |           |             |          |          |

## 统计决策

对于列因素A，F的检验统计值大于临界值，fa=3.830769>3.259167,所以因拒绝H0
对于行因素B，F的检验统计值小于临界值，fb=1.897436<3.490295,所以因接受H0

# 例题8.1

In [8]:
data = pd.read_excel(r"C:\Users\柠檬\Desktop\实验5\指数-例8-1.xlsx",header=[0,1])
data=data.iloc[:3,:6]
data

商品种类               计量单位     销售量             价格       
  Unnamed: 0_level_1 Unnamed: 1_level_1   9月 q0  10月 q1  9月 p0 10月 p1
0                  甲                  台   220.0   340.0  750.0  680.0
1                  乙                  袋  3400.0  4270.0  150.0  120.0
2                  丙                  件   141.0   199.0  328.0  286.0

## 拉式销售量综合指数

In [9]:
Iq=0
x=0
y=0
for i in range(3):
    x+=data.values[i,4]*data.values[i,3]
    y+=data.values[i,4]*data.values[i,2]
Iq=x/y
Iq

1.3320965881361195

与9月相比，10月3种商品销售量总体而言增加了33.21%

## 帕氏价格综合指数

In [10]:
data.values[:,2]

array([220.0, 3400.0, 141.0], dtype=object)

In [11]:
Ip=0
x=0
y=0
for i in range(3):
    x+=data.values[i,5]*data.values[i,3]
    y+=data.values[i,4]*data.values[i,3]
Ip=x/y
Ip

0.8331987193631788

与9月相比，10月3种商品价格总体而言下降了了16.68%

# 例题8.2+8.3

## 读取数据

In [12]:
data1 = pd.read_excel(r"C:\Users\柠檬\Desktop\实验5\指数-例8-2.xlsx")
drop_row = list(range(4,9))
drop_row.append(0)
data1 = data1.drop(drop_row)
drop_col = ["Unnamed: {}".format(_) for _ in range(4,10)]
drop_col.append('计量单位')
data1 = data1.drop(drop_col,axis=1)
data1 = data1.iloc[:,[0,1,2]]
data1.columns = ['商品种类','q0','q1']
display(data1)

商品种类    q0    q1
1    甲   220   340
2    乙  3400  4270
3    丙   141   199

In [13]:
data2 = pd.read_excel(r"C:\Users\柠檬\Desktop\实验5\指数-例8-3.xlsx")
drop_row = list(range(4,10))
drop_row.append(0)
data2 = data2.drop(drop_row)
drop_col = ["Unnamed: {}".format(_) for _ in range(4,10)]
drop_col.append('计量单位')
data2 = data2.drop(drop_col,axis=1)
data2 = data2.iloc[:,[0,1,2]]
data2.columns = ['商品种类','p0','p1']
display(data2)

商品种类   p0   p1
1    甲  750  680
2    乙  150  120
3    丙  328  286

## 计算计算机甲、乙、丙3种商品的销售量总指数

In [14]:
Iq = sum([kq * p0 * q0 for kq,p0,q0 in zip(data1.iloc[:,2] / data1.iloc[:,1],data2.iloc[:,1],data1.iloc[:,1])])/sum([p0 * q0 for p0,q0 in zip(data2.iloc[:,1],data1.iloc[:,1])])
print("总指数：%.2f"%(Iq*100)+"%")

总指数：133.21%


## 计算机甲、乙、丙3种商品的价格总指数。

In [15]:
Ip = sum([p1 * q1 for p1,q1 in zip(data2.iloc[:,2],data1.iloc[:,2])])/sum([(1 / kp) * p1 * q1 for kp,p1,q1 in zip(data2.iloc[:,2] / data2.iloc[:,1],data2.iloc[:,2],data1.iloc[:,2])])
print("总指数：%.2f"%(Ip*100)+"%")

总指数：83.32%


# Makeup2007

In [16]:
experiment = pd.read_excel(r"C:\Users\柠檬\Desktop\实验5\Makeup2007.xls",skiprows=2)
experiment = experiment.drop(["Unnamed: {}".format(_) for _ in (0,1)],axis=1)
date = sorted(set(experiment['Date']))
left = date[:len(date)//2]
def date2mouth(x):
    if x in left:
        return "front"
    else:
        return "after"
experiment['时间'] = experiment['Date'].apply(date2mouth)
ex = experiment.drop(['Transaction number','Name','Date','Location'],axis=1)
pt = pd.pivot_table(ex,values=['Units','Dollars'],index=['Product'],columns=['时间'],aggfunc='sum')
ex

Product  Units     Dollars     时间
0      lip gloss     45  137.204558  front
1     foundation     50  152.007303  front
2       lipstick      9   28.719483  front
3      lip gloss     55  167.075323  after
4      lip gloss     43  130.602872  front
...          ...    ...         ...    ...
1699  foundation     36  109.842599  front
1700   lip gloss     46  140.408899  after
1701    lipstick     72  217.835886  front
1702   eye liner     28   85.656830  after
1703   eye liner     54  164.487334  after

[1704 rows x 4 columns]

## 拉氏销售量综合指数

In [17]:
Iq = sum([p0 * q1 for p0,q1 in zip(pt['Dollars']['front'],pt['Units']['after'])]) / sum([p0 * q0 for p0,q0 in zip(pt['Dollars']['front'],pt['Units']['front'])])
print("拉氏销售量综合指数：{:.2f}%".format(Iq*100))

拉氏销售量综合指数：95.94%


## 帕氏销售量综合指数

In [18]:
Iq = sum([p1 * q1 for p1,q1 in zip(pt['Dollars']['after'],pt['Units']['after'])]) / sum([p1 * q0 for p1,q0 in zip(pt['Dollars']['after'],pt['Units']['front'])])
print("帕氏销售量综合指数：{:.2f}%".format(Iq*100))

帕氏销售量综合指数：96.22%


## 加权算数平均指数法

In [19]:
Iq = sum([kq * p0 * q0 for kq,p0,q0 in zip(pt['Units']['after'] / pt['Units']['front'],pt['Dollars']['front'],pt['Units']['front'])])\
    /sum([p0 * q0 for p0,q0 in zip(pt['Dollars']['front'],pt['Units']['front'])])
print("销售量总指数：{:.2f}%".format(Iq*100))

销售量总指数：95.94%


## 以加权调和平均指数法

In [20]:
Ip = sum([p1 * q1 for p1,q1 in zip(pt['Dollars']['after'],pt['Units']['after'])])\
    /sum([(1 / kp) * p1 * q1 for kp,p1,q1 in zip(pt['Dollars']['after'] / pt['Dollars']['front'],pt['Dollars']['after'],pt['Units']['after'])])
print("价格总指数：{:.2f}%".format(Ip*100))

价格总指数：96.17%
